langkit/ocaml_api.py

```diff
@@ -0,0 +1,444 @@
+    """Container for OCaml API generation settings."""
+
+    name = 'ocaml'
+
+    def __init__(self, ctx, c_api_settings):
+        self.context = ctx
+        self.c_api_settings = c_api_settings
+
+    def is_empty_type(self, type):
+        """
+        Test if the given type would end-up being empty, so that we can avoid
+        generating empty types.
+
+        :param ct.CompiledType: The type we want to check if it is empty.
+        :rtype: bool
+        """
+        return dispatch_on_type(type, [
+            (T.Bool, lambda _: False),
+            (T.Int, lambda _: False),
+            (T.Character, lambda _: False),
+            (T.Token, lambda t: False),
+            (T.Symbol, lambda _: False),
+            (ct.EnumType, lambda t: False),
+            (ct.ASTNodeType, lambda t:
+                len(t.concrete_subclasses) == 0),
+            (ct.EntityType, lambda t: self.is_empty_type(t.astnode)),
+            (T.AnalysisUnit, lambda _: False),
+            (ct.ArrayType, lambda t: self.is_empty_type(t.element_type)),
+            (ct.StructType, lambda t:
+                len([
+                    t
+                    for field in t.get_fields()
+                    if not self.is_empty_type(field.type)
+                ]) == 0),
+            (T.BigInt, lambda _: False),
+            (T.EnvRebindings, lambda _: False),
+        ])
+
+    def polymorphic_variant_name(self, type):
+        """
+        Return a string representing the polymorphic variant for the given
+        concrete type.
+
+        :param ct.ASTNodeType type: The type for which we want the
+            polymorphic variant name.
+        :rtype: list[ct.CompiledType]
+        """
+        return "`{}".format(type.entity.api_name.camel)
+
+    def fields_name(self, for_module, from_module=None):
+        """
+        Return a string representing the name of the ocaml type for the
+        fields of a node.
+
+        :param ct.ASTNodeType for_module: The field comes from this module.
+        :param ct.ASTNodeType from_module: The field is accessed inside this
+            module.
+        :rtype: list[ct.CompiledType]
+        """
+        if for_module == from_module:
+            return 'fields'
+        else:
+            return "{}.{}".format(self.module_name(for_module), 'fields')
+
+    def check_for_null(self, c_value, astnode, context):
+        """
+        Return an ocaml expression that first check if the underlying node of
+        c_value is null. If it is the case, the expression is evaluated to
+        None, otherwise, wrap c_value to the given astnode type.
+
+        :param str c_value: The ocaml c value for a node.
+        :param ct.ASTNodeType: The type of c_value.
+        :param str context: The ocaml expression denoting the context.
+        """
+
+        return (
+            "if is_null (getf {} EntityStruct.node) then None else Some ({})".
+            format(c_value, self.wrap_value(c_value, astnode, context))
```

Comment:
In this case we generally put the dot at the beginning of the next line: could you move it?